### PR TITLE
node: nodes should fetch state on startup

### PR DESCRIPTION
**Pull request by tychoish**
This is in persuit of #6777, but sort of backs into the solution, by
removing the construction-time dependency on a state object and uses
the state store and reorganizes reactor construction to interact with
the during startup.

### CODE REVIEW

File: internal/state/execution.go [hunk 1]
```diff
@@ -324,8 +324,8 @@ func (blockExec *BlockExecutor) VerifyVoteExtension(ctx context.Context, vote *t
 // state before new txs are run in the mempool, lest they be invalid.
 func (blockExec *BlockExecutor) Commit(
 	ctx context.Context,
-	state State,
 	block *types.Block,
+	state State,
```
Comment (williambanfield) [inline internal/state/execution.go:328]: Why this change?

File: node/setup.go [hunk 2]
```diff
@@ -184,10 +184,9 @@ func createMempoolReactor(
 		logger,
 		cfg.Mempool,
 		proxyApp.Mempool(),
-		state.LastBlockHeight,
 		mempool.WithMetrics(memplMetrics),
-		mempool.WithPreCheck(sm.TxPreCheck(state)),
-		mempool.WithPostCheck(sm.TxPostCheck(state)),
+		mempool.WithPreCheck(sm.TxPreCheckFromStore(store)),
+		mempool.WithPostCheck(sm.TxPostCheckFromStore(store)),
```
Comment (williambanfield) [inline node/setup.go:189]: Does the old way of doing this mean that the checks did not change over the lifetime of the node, even if the values in the state store changed? That seems kind of incorrect, so I'm glad we're fixing it.

File: internal/mempool/mempool.go [hunk 3]
```diff
@@ -94,15 +94,14 @@ func NewTxMempool(
...
-	height int64,
 	options ...TxMempoolOption,
 ) *TxMempool {
 
 	txmp := &TxMempool{
 		logger:        logger,
 		config:        cfg,
 		proxyAppConn:  proxyAppConn,
-		height:        height,
+		height:        -1,
```
Comment (williambanfield) [inline internal/mempool/mempool.go:104]: Probably missing it, but where is this set now?

File: internal/state/tx_filter.go [hunk 4]
```diff
@@ -1,22 +1,85 @@
...
 	"github.com/tendermint/tendermint/types"
 )
 
-// TxPreCheck returns a function to filter transactions before processing.
-// The function limits the size of a transaction to the block's maximum data size.
-func TxPreCheck(state State) mempool.PreCheckFunc {
-	maxDataBytes := types.MaxDataBytesNoEvidence(
-		state.ConsensusParams.Block.MaxBytes,
-		state.Validators.Size(),
+func cachingStateFetcher(store Store) func() (State, error) {
```
Comment (williambanfield) [inline internal/state/tx_filter.go:12]: This is an improvement since it looks like we didn't update the check in the mempool at all in the previous code. I do think a height-based approach would be a bit better. We could update the `PostCheckTx` and `PreCheckTx` methods to take the height of the chain and have the caching fetcher update when it sees a new height.

File: internal/consensus/state.go [hunk 5]
```diff
@@ -365,6 +360,18 @@ func (cs *State) LoadCommit(height int64) *types.Commit {
 // OnStart loads the latest state via the WAL, and starts the timeout and
 // receive routines.
 func (cs *State) OnStart(ctx context.Context) error {
+	state, err := cs.stateStore.Load()
+	if err != nil {
+		return err
```
Comment (creachadair) [inline internal/consensus/state.go:365]:
```suggestion
		return fmt.Errorf("loading state: %w", err)
```

**Review by creachadair — APPROVED** [review]

File: internal/mempool/mempool.go (hunk 3 above)
Comment (tychoish) [reply internal/mempool/mempool.go:104]: This is only used in a logging statement and it's updated in `Update()`

File: internal/state/execution.go (hunk 1 above)
Comment (tychoish) [reply internal/state/execution.go:328]: oops, fixking

File: internal/state/tx_filter.go (hunk 4 above)
Comment (tychoish) [reply internal/state/tx_filter.go:12]: the Update method more or less does this, though the state store object doesn't allow you to access previous states so the actual final design here is something we should think more about. 

I think in general we should pass a store around in places where we currently pass a state (particularly during construction/startup, to avoid needing to connect to the app until we start (which is the core of #6777), so that's what this change does.

Having said that, the cache is probably a bit over wrought, but I think calling Load() for every single tx is probably also likely to cause a problem. And the case where Update isn't called on the mempool wouldn't behave meaingfully worse than the current implementation.

**Review by williambanfield — APPROVED** [review]

**Comment by cmwaters** [comment]
As a slightly alternative proposal what if the `BlockExecutor` was the only struct to contain the `Statestore` and `Blockstore` and all "state advancing protocols" i.e. `Consensus`, `Blocksync` and `Statesync` relied on the `BlockExecutor` to make updates to `State` or to "execute" a `Block`. They couldn't update the stores directly.

**Comment by tychoish** [comment]
> As a slightly alternative proposal what if the `BlockExecutor` was the only struct to contain the `Statestore` and `Blockstore` and all "state advancing protocols" i.e. `Consensus`, `Blocksync` and `Statesync` relied on the `BlockExecutor` to make updates to `State` or to "execute" a `Block`. They couldn't update the stores directly.

I think that's kind of orthogonal to this change and definitely something we could (and should?) work towards. Most of these cases are read-only uses where we're using the store object (like a database) to look at the state which becomes available after the node is started, but we have to give (various components) access to the state (for consensus parameters or similar,) when construction happens. 

If we put the state store/block store behind (or in) another interface we'd still have to pass it to all of the relevant components (which is fine).
